src/librustc_mir/borrow_check/mod.rs

```diff
@@ -992,36 +1009,39 @@ impl<'cx, 'gcx, 'tcx> MirBorrowckCtxt<'cx, 'gcx, 'tcx> {
                 self.visit_terminator_drop(
                     loc, term, flow_state, &deref_place, ty, span, seen);
 
-                // Second, we model the drop of the box-pointer itself
-                // via a shallow write.
+                // Second, we model the deallocation of the box
+                // backing storage via a *shallow* StorageDeadOrDrop
+                // write to a *deref* of the box. Note that the
+                // StorageDead of the box-pointer below in step 3
+                // wouldn't do such a deref.
                 //
-                // FIXME: Could we just get by with the last step
-                // below? (This second step seems like a good thing to
-                // do just as a matter of principle; but none of our
-                // testsuite complained when I tried leaving it out.)
+                // (We use a shallow write here because a Deep one
+                // would touch `&mut` references in `T`, which would
+                // defeat the purpose of this special case code.)
                 self.access_place(
                     ContextKind::Drop.new(loc),
-                    (drop_place, span),
+                    (&deref_place, span),
                     (ShallowOrDeep::Shallow(None), Write(WriteKind::StorageDeadOrDrop)),
                     LocalMutationIsAllowed::Yes,
                     flow_state,
                 );
 
-                // Third, we model the deallocation of the box
-                // contents via a shallow write to a *deref* of the
-                // box, because the shallow write above wouldn't do
-                // such a deref.
+                // Third, we model the StorageDead of the box-pointer
+                // representation itself via a shallow write.
                 //
-                // (We use a shallow write because a Deep one would
-                // touch `&mut` references in `T`.)
+                // FIXME: Could we just get by with just steps 1 and 2
+                // above? (This last step seems like a good thing to
+                // do just as a matter of principle; but none of our
+                // testsuite complained when I tried leaving it out.)
```

Comment:
I think it's relevant if part of liballoc did `&a_box.0` (for whatever reason) - without this last step, `a_box` could be dropped without that borrow being invalidated, right? You could probably use a `#![no_std]` test to check for this, by definition your own `Box` lang item.